ports/stm32/boards/OPENMV_N6/board.c

```diff
@@ -0,0 +1,261 @@
+            | 1 << XSPI_CCR_DMODE_Pos // data on 1 line
+            | 0 << XSPI_CCR_ABMODE_Pos // no alternate byte
+            | 0 << XSPI_CCR_ADMODE_Pos // no address
+            | 1 << XSPI_CCR_IMODE_Pos // instruction on 1 line
+    ;
+
+    XSPI2->TCR = 0 << XSPI_TCR_DCYC_Pos; // 0 dummy cycles
+
+    // This triggers the start of the operation.
+    XSPI2->IR = cmd; // read opcode
+
+    // Read in remaining data 1 byte at a time
+    while (len) {
+        while (!((XSPI2->SR >> XSPI_SR_FLEVEL_Pos) & 0x3f)) {
+            if (XSPI2->SR & XSPI_SR_TEF) {
+                return -MP_EIO;
+            }
+        }
+        *dest++ = *(volatile uint8_t *)&XSPI2->DR;
+        --len;
+    }
+
+    XSPI2->FCR = XSPI_FCR_CTCF; // clear TC flag
+
+    return 0;
+}
+
+static void xspi_memory_map_111(void) {
+    XSPI2->CCR =
+        1 << XSPI_CCR_DMODE_Pos // data on 1 line
+            | 3 << XSPI_CCR_ADSIZE_Pos // 32-bit address
+            | 1 << XSPI_CCR_ADMODE_Pos // address on 1 line
+            | 1 << XSPI_CCR_IMODE_Pos // instruction on 1 line
+    ;
+
+    XSPI2->TCR = 0 << XSPI_TCR_DCYC_Pos; // no dummy cycles
+    XSPI2->IR = 0x13; // READ4B
+    XSPI2->LPTR = 1024; // timeout period in number of CLK cycles
+
+    // Enable the XSPI peripheral in memory-mapped mode.
+    XSPI2->CR = (XSPI2->CR & ~XSPI_CR_FMODE_Msk) | 3 << XSPI_CR_FMODE_Pos;
+
+    // TODO: enable MPU?
+}
+
+static void xspi_memory_map_888(void) {
+    XSPI2->CCR =
+        1 << XSPI_CCR_DQSE_Pos // DQS enabled
+            | 1 << XSPI_CCR_DDTR_Pos // data DTR enabled
+            | 4 << XSPI_CCR_DMODE_Pos // data on 8 lines
+            | 0 << XSPI_CCR_ABSIZE_Pos // 8-bit alternate bytes
+            | 0 << XSPI_CCR_ABDTR_Pos // alternate bytes DTR disabled
+            | 0 << XSPI_CCR_ABMODE_Pos // no alternate bytes
+            | 3 << XSPI_CCR_ADSIZE_Pos // 32-bit address
+            | 1 << XSPI_CCR_ADDTR_Pos // address DTR enabled
+            | 4 << XSPI_CCR_ADMODE_Pos // address on 8 lines
+            | 1 << XSPI_CCR_ISIZE_Pos // 16-bit instruction
+            | 1 << XSPI_CCR_IDTR_Pos // instruction DTR enabled
+            | 4 << XSPI_CCR_IMODE_Pos // instruction on 8 lines
+    ;
+
+    XSPI2->TCR =
+        0 << XSPI_TCR_SSHIFT_Pos // no sample shift
+            //| 1 << XSPI_TCR_DHQC_Pos // 1/4 cycle hold enabled
+            | 6 << XSPI_TCR_DCYC_Pos // 6 dummy cycles for reading (minimum, flash may insert more by holding DQS low)
+    ;
+
+    XSPI2->IR = 0xee11; // OCTA DTR read mode (8DTRD)
+
+    XSPI2->LPTR = 1024; // timeout period in number of CLK cycles
+
+    // Enable the XSPI peripheral in memory-mapped mode.
+    XSPI2->CR = (XSPI2->CR & ~XSPI_CR_FMODE_Msk) | 3 << XSPI_CR_FMODE_Pos;  // memory-mapped mode
+
+    // TODO: enable MPU?
+}
+
+#if BUILDING_MBOOT
+
+#include "mboot/mboot.h"
+
+#else
+
+void board_early_init(void) {
+    // TODO: if (HAL_PWREx_ConfigSupply(PWR_EXTERNAL_SOURCE_SUPPLY ) != HAL_OK)
+    //xspi_flash_init();
+}
+
```

Comment:
```suggestion

#define OMV_BOOT_MAGIC_ADDR   (0x3401FFFCU)
#define OMV_BOOT_MAGIC_VALUE  (0xB00710ADU)

void board_enter_bootloader(void) {
    *((uint32_t *) OMV_BOOT_MAGIC_ADDR) = OMV_BOOT_MAGIC_VALUE;
    SCB_CleanDCache();
    NVIC_SystemReset();
}
```
To enter our bootloader.

Reply:
Do you want to have the OPENMV_N6 board definition included in MicroPython's stm32 port (like with the OPENMV_AE3)?  If so, it should work with mboot.  But probably also a good idea to work with the OpenMV bootloader, which means adding this code.

Related: how did you choose address 0x3401FFFC?  That's part way through SRAM1, in the FLEXRAM area.  You'd need to make sure that isn't cleared on reset like the rest of SRAM1/2, and also make sure it's not overwritten by the bootloader.

Reply:
> Do you want to have the OPENMV_N6 board definition included in MicroPython's stm32 port (like with the OPENMV_AE3)? If so, it should work with mboot. But probably also a good idea to work with the OpenMV bootloader, which means adding this code.

Either way is fine with me. If mboot support is kept, I can always use `MP_CONFIGFILE` to override and not build mboot when building our firmware.

> Related: how did you choose address 0x3401FFFC? That's part way through SRAM1, in the FLEXRAM area. You'd need to make sure that isn't cleared on reset like the rest of SRAM1/2, and also make sure it's not overwritten by the bootloader.

Good point. The bootloader uses the 128K (or 64K) DTCM for its memory (heap, stack etc..), so this address is typically `0x2001FFFCU` (last word of bootloader's memory) and it's the same for almost all boards. However, for some reason, this doesn't work on the N6, so I just used SRAM1. Note that SRAM1 and SRAM2 don't seem to get erased on reset, otherwise this wouldn't work, but it would still be better to use DTCM for this, but it's not working.

Reply:
Oh I think I just needed a DSB, seems the write might be buffered. The following works too:

```C
#define OMV_BOOT_MAGIC_ADDR   (0x3001FFFCU)
#define OMV_BOOT_MAGIC_VALUE  (0xB00710ADU)

void board_enter_bootloader(void) {
    *((uint32_t *) OMV_BOOT_MAGIC_ADDR) = OMV_BOOT_MAGIC_VALUE;
    __DSB();
    NVIC_SystemReset();
}
```

Reply:
@dpgeorge There seems to be some issue with DTCM, I can't access any address above ~0x1000. This causes a fault on boot when it tries accesses the magic number address (oddly enough, it doesn't always happen). Perhaps it has something to do with security config, flexram, clocks or something else that gets enabled by the main firmware, but even from the main firmware I still can't access this memory from gdb.  Anyway let's please keep the original boot address (with cache invalidate).